src/core/net/dns_headers.hpp

```diff
@@ -788,9 +938,18 @@ class Name
         uint16_t       mNameEndOffset;    // Offset in `mMessage` to the byte after the end of domain name field.
     };
 
-    Name(void) = default;
-
     static otError AppendLabel(const char *aLabel, uint8_t aLabelLength, Message &aMessage);
+
+    Name(const char *aString, const Message *aMessage, uint16_t aOffset)
+        : mString(aString)
+        , mMessage(aMessage)
+        , mOffset(aOffset)
+    {
+    }
+
+    const char *   mString;  // String containing the name or `nullptr` if name is not from string.
+    const Message *mMessage; // Message containing the encoded name, or `nullptr` if `Name` is not from message.
+    uint16_t       mOffset;  // Offset in `mMessage` to the start of name (used when name is from `mMessage`).
```

Comment:
```suggestion
   union {
      const char *   mString;  // String containing the name or `nullptr` if name is not from string.     
      struct {
        const Message *mMessage; // Message containing the encoded name, or `nullptr` if `Name` is not from message.
        uint16_t       mOffset;  // Offset in `mMessage` to the start of name (used when name is from `mMessage`).
      } 
   }
```

Use union to reduce memory? We can use `mOffset` to determine whether it's a string (mOffset == 0) or not (mOffset > 0) because `mOffset` should never be zero when it's encoded.

Reply:
The idea of using `mOffset` of zero to indicate the type is interesting and smart, but I am not sure if we should go for it since I think it may be limiting. It's correct that in a DNS message the offset should never be zero, but we may use a `Message` as a data blob storage, for example to store metadata about query (we do this actually in the new DNS client) and in that case it can happen that a name is encoded at any offset (maybe at offset zero).

-------

On using `union`, I actually started with that. Basically I had a `mType` and then a `union` but decided against it.

- On the upside, I agree it can reduce the `sizeof(Name)` by few bytes. But I think in all cases, the instances of `Name` would be temporary/local vairable and the size won't matter much.
- On the downside, it adds a code size and computation overhead (since now we need to check `mType` against set of values). 

Not using `union` has the advantage that complier/toolchain can optimize/inline the code better:
- Compliers typically generate very efficient code fo comparing a pointer/value against zero (this can end up being combined into the instruction that loads the value from memory to a register (same instruction does zero comparison as well).
- Most advanced toolchain can optimize simple struct instance when passing them to methods/function and use  registers to pass the fields rather than allocating stack space to generate the object. `union` won't be optimized like this.

Reply:
I am fine to keep it as is since Name is mostly allocated on Heap.